### PR TITLE
sys/base64: Clean up / improve implementation

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

Improves the implementation to be (subjectively) more readable and (objectively) perform better.

### Testing procedure

1. `make -C tests/unittests/ BOARD=<SOME_BOARD> tests-base64 flash test`
2. `make -C tests/bench_sys_base64 BOARD=<SOME_BOARD> flash term` (for both this PR, and for #14400)

### Issues/PRs references

Depends on and include #14400. Only the lasted commit belongs to this PR.

### CODE REVIEW

**Comment by maribu** [comment]
Here are a bunch of tests done with this PR compared to the performance in #14400. (Performance-wise, #14400 and master should be identical.) I ran all tests after `export BUILD_IN_DOCKER=1`, so that other can easily reproduce.

**Note:** The branch name of this PR is two chars longer, which results in the message `main(): This is RIOT! (Version: ...)` consuming two bytes more. To compensate, I subtracted two bytes from the `.text` section of this PR. For the Arduino Nano (AVR is a Harvard arch) the message is stored in `.data` instead. For the EPS32, I did no compensation, as I don't know the architecture well enough; e.g. there might by some padding / alignment stuff going on I'm not aware of.

#### `nucleo-f767zi`

ROM:            -14 Byte (see note)
RAM:            +/-0
Time encode:    -25.0%
Time decode:    -32.1%

<details><summary>https://github.com/RIOT-OS/RIOT/pull/14400</summary>

```
   text	   data	    bss	    dec	    hex	filename
  11144	    112	   2484	  13740	   35ac	/data/riotbuild/riotbase/tests/bench_sys_base64/bin/nucleo-f767zi/tests_bench_sys_base64.elf
```

```
2020-07-02 07:52:18,993 # main(): This is RIOT! (Version: 2020.07-devel-1673-gebba6-base64_tests)
2020-07-02 07:52:18,999 # Verifying that base64 encoding works for benchmark input: OK
2020-07-02 07:52:19,004 # Verifying that base64 decoding works for benchmark input: OK
2020-07-02 07:52:19,765 # Encoding 10.000 x 96 bytes (128 bytes in base64): 755743 µs
2020-07-02 07:52:20,818 # Decoding 10.000 x 96 bytes (128 bytes in base64): 1047316 µs
```

</details>

<details><summary>This PR</summary>

```
   text	   data	    bss	    dec	    hex	filename
  11132	    112	   2484	  13728	   35a0	/data/riotbuild/riotbase/tests/bench_sys_base64/bin/nucleo-f767zi/tests_bench_sys_base64.elf
```

```
2020-07-02 09:02:56,423 # main(): This is RIOT! (Version: 2020.07-devel-1672-gcc3be-base64_cleanup)
2020-07-02 09:02:56,430 # Verifying that base64 encoding works for benchmark input: OK
2020-07-02 09:02:56,434 # Verifying that base64 decoding works for benchmark input: OK
2020-07-02 09:02:57,006 # Encoding 10.000 x 96 bytes (128 bytes in base64): 566808 µs
2020-07-02 09:02:57,723 # Decoding 10.000 x 96 bytes (128 bytes in base64): 711437 µs
```

</details>

#### `nucleo-f446re`

ROM:            -30 Byte (see note)
RAM:            +/-0
Time encode:    -40.2%
Time decode:    -19.6%

<details><summary>https://github.com/RIOT-OS/RIOT/pull/14400</summary>

```
   text	   data	    bss	    dec	    hex	filename
  11064	    112	   2484	  13660	   355c	/data/riotbuild/riotbase/tests/bench_sys_base64/bin/nucleo-f446re/tests_bench_sys_base64.elf
```

```
2020-07-02 07:53:40,474 # main(): This is RIOT! (Version: 2020.07-devel-1673-gebba6-base64_tests)
2020-07-02 07:53:40,479 # Verifying that base64 encoding works for benchmark input: OK
2020-07-02 07:53:40,484 # Verifying that base64 decoding works for benchmark input: OK
2020-07-02 07:53:40,784 # Encoding 10.000 x 96 bytes (128 bytes in base64): 293890 µs
2020-07-02 07:53:41,222 # Decoding 10.000 x 96 bytes (128 bytes in base64): 432557 µs
```

</details>

<details><summary>This PR</summary>

```
   text	   data	    bss	    dec	    hex	filename
  11036	    112	   2484	  13632	   3540	/data/riotbuild/riotbase/tests/bench_sys_base64/bin/nucleo-f446re/tests_bench_sys_base64.elf

```

```
2020-07-02 09:04:46,462 # main(): This is RIOT! (Version: 2020.07-devel-1672-gcc3be-base64_cleanup)
2020-07-02 09:04:46,467 # Verifying that base64 encoding works for benchmark input: OK
2020-07-02 09:04:46,473 # Verifying that base64 decoding works for benchmark input: OK
2020-07-02 09:04:46,654 # Encoding 10.000 x 96 bytes (128 bytes in base64): 175834 µs
2020-07-02 09:04:47,007 # Decoding 10.000 x 96 bytes (128 bytes in base64): 347668 µs
```

</details>

#### `wemos-zero`

ROM:            -18 Byte (see note)
RAM:            +/-0
Time encode:    -47.5%
Time decode:    -29.1%

<details><summary>https://github.com/RIOT-OS/RIOT/pull/14400</summary>

```
   text	   data	    bss	    dec	    hex	filename
  17976	    132	   6044	  24152	   5e58	/data/riotbuild/riotbase/tests/bench_sys_base64/bin/wemos-zero/tests_bench_sys_base64.elf
```

```
2020-07-02 07:55:11,526 # main(): This is RIOT! (Version: 2020.07-devel-1673-gebba6-base64_tests)
2020-07-02 07:55:11,529 # Verifying that base64 encoding works for benchmark input: OK
2020-07-02 07:55:11,532 # Verifying that base64 decoding works for benchmark input: OK
2020-07-02 07:55:11,533 # Encoding 10.000 x 96 bytes (128 bytes in base64): 2137446 µs
2020-07-02 07:55:14,628 # Decoding 10.000 x 96 bytes (128 bytes in base64): 3132327 µs
```

</details>

<details><summary>This PR</summary>

```
   text	   data	    bss	    dec	    hex	filename
  17960	    132	   6044	  24136	   5e48	/data/riotbuild/riotbase/tests/bench_sys_base64/bin/wemos-zero/tests_bench_sys_base64.elf
```

```
(Note: Terminal connected to late to capture `main(): This is RIOT!...` message.)

2020-07-02 09:07:23,871 # Verifying that base64 encoding works for benchmark input: OK
2020-07-02 09:07:23,873 # Verifying that base64 decoding works for benchmark input: OK
2020-07-02 09:07:23,900 # Encoding 10.000 x 96 bytes (128 bytes in base64): 1122654 µs
2020-07-02 09:07:26,127 # Decoding 10.000 x 96 bytes (128 bytes in base64): 2219827 µs
```

</details>

#### `esp32-mh-et-live-minikit`

ROM:            -24 Byte (no compensation for longer branch name here, see note)
RAM:            +/-0
Time encode:    -11.2%
Time decode:    -28.4%

<details><summary>https://github.com/RIOT-OS/RIOT/pull/14400</summary>

```
   text	   data	    bss	    dec	    hex	filename
  57227	   3628	   6512	  67367	  10727	/data/riotbuild/riotbase/tests/bench_sys_base64/bin/esp32-mh-et-live-minikit/tests_bench_sys_base64.elf
```

```
2020-07-02 07:57:53,572 # main(): This is RIOT! (Version: 2020.07-devel-1673-gebba6-base64_tests)
2020-07-02 07:57:53,575 # Verifying that base64 encoding works for benchmark input: OK
2020-07-02 07:57:53,576 # Verifying that base64 decoding works for benchmark input: OK
2020-07-02 07:57:53,576 # Encoding 10.000 x 96 bytes (128 bytes in base64): 709006 µs
2020-07-02 07:57:54,831 # Decoding 10.000 x 96 bytes (128 bytes in base64): 1519258 µs
```

</details>

<details><summary>This PR</summary>

```
   text	   data	    bss	    dec	    hex	filename
  57203	   3628	   6512	  67343	  1070f	/data/riotbuild/riotbase/tests/bench_sys_base64/bin/esp32-mh-et-live-minikit/tests_bench_sys_base64.elf

```

```
2020-07-02 09:09:38,001 # main(): This is RIOT! (Version: 2020.07-devel-1672-gcc3be-base64_cleanup)
2020-07-02 09:09:38,003 # Verifying that base64 encoding works for benchmark input: OK
2020-07-02 09:09:38,004 # Verifying that base64 decoding works for benchmark input: OK
2020-07-02 09:09:38,005 # Encoding 10.000 x 96 bytes (128 bytes in base64): 629256 µs
2020-07-02 09:09:38,757 # Decoding 10.000 x 96 bytes (128 bytes in base64): 1088127 µs
```

</details>

#### `arduino-nano`

ROM:            -156 Byte
RAM:            +/-0 (see note)
Time encode:    -65.4%
Time decode:    -54.8%

<details><summary>https://github.com/RIOT-OS/RIOT/pull/14400</summary>

```
   text	   data	    bss	    dec	    hex	filename
   8878	    708	   1042	  10628	   2984	/data/riotbuild/riotbase/tests/bench_sys_base64/bin/arduino-nano/tests_bench_sys_base64.elf
```

```
2020-07-02 08:01:51,292 # main(): This is RIOT! (Version: 2020.07-devel-1673-gebba6-base64_tests)
2020-07-02 08:01:51,359 # Verifying that base64 encoding works for benchmark input: OK
2020-07-02 08:01:51,398 # Verifying that base64 decoding works for benchmark input: OK
2020-07-02 08:02:01,056 # Encoding 10.000 x 96 bytes (128 bytes in base64): 9589936 µs
2020-07-02 08:02:13,868 # Decoding 10.000 x 96 bytes (128 bytes in base64): 12742804 µs
```

</details>

<details><summary>This PR</summary>

```
   text	   data	    bss	    dec	    hex	filename
   8722	    710	   1042	  10474	   28ea	/data/riotbuild/riotbase/tests/bench_sys_base64/bin/arduino-nano/tests_bench_sys_base64.elf
```


```
2020-07-02 09:12:43,311 # main(): This is RIOT! (Version: 2020.07-devel-1672-gcc3be-base64_cleanup)
2020-07-02 09:12:43,377 # Verifying that base64 encoding works for benchmark input: OK
2020-07-02 09:12:43,419 # Verifying that base64 decoding works for benchmark input: OK
2020-07-02 09:12:46,800 # Encoding 10.000 x 96 bytes (128 bytes in base64): 3316708 µs
2020-07-02 09:12:52,631 # Decoding 10.000 x 96 bytes (128 bytes in base64): 5766112 µs
```
</details>

**Comment by aabadie** [comment]
> The branch name of this PR is two chars longer, which results in the message

You can also use `RIOT_VERSION=test` to have consistent results between master and a PR.

**Comment by maribu** [comment]
> You can also use `RIOT_VERSION=test` to have consistent results between master and a PR.

Thanks for pointing this out. Next time I'll do so :-)

**Comment by fjmolinas** [comment]
This one needs rebase now

File: sys/base64/base64.c [hunk 1]
```diff
@@ -232,34 +229,44 @@ int base64_decode(const void *base64_in, size_t base64_in_size,
...
+                    case 3:
+                        codes[3] = 0;
+                        decode_four_codes(out, codes);
+                        out += 2;
+                        break;
+                }
+                *data_out_size = (size_t)out - (size_t)data_out;
+                return BASE64_SUCCESS;
+            }
+            switch (codes[n_codes] = getcode(*in++)) {
```
Comment (benpicco) [inline sys/base64/base64.c:260]:
```suggestion
            /* load up to 4 codes into decode buffer, skip invalid symbols */
            switch (codes[n_codes] = getcode(*in++)) {
```

File: sys/base64/base64.c [hunk 2]
```diff
@@ -232,34 +229,44 @@ int base64_decode(const void *base64_in, size_t base64_in_size,
...
-        if (code == BASE64_NOT_DEFINED || code == BASE64_EQUALS) {
-            continue;
+    const uint8_t *end = base64_in + base64_in_size;
+    uint8_t codes[4];
+    size_t n_codes;
+
+    while (1) {
+        n_codes = 0;
+        do {
+            if (in == end) {
```
Comment (benpicco) [inline sys/base64/base64.c:239]:
```suggestion
            /* decode last remaining codes */
            if (in == end) {
```

File: sys/base64/base64.c [hunk 3]
```diff
@@ -232,34 +229,44 @@ int base64_decode(const void *base64_in, size_t base64_in_size,
...
-    int code = 0;
-
-    int mask = 2;
-
-    for (int i = 1; i < (int)(base64_in_size); i++) {
-        code = getcode(in[i]);
-
-        if (code == BASE64_NOT_DEFINED || code == BASE64_EQUALS) {
-            continue;
+    const uint8_t *end = base64_in + base64_in_size;
```
Comment (benpicco) [inline sys/base64/base64.c:232]:
```suggestion
    const uint8_t *end = in + base64_in_size;
```

File: sys/base64/base64.c [hunk 4]
```diff
@@ -232,34 +229,44 @@ int base64_decode(const void *base64_in, size_t base64_in_size,
...
+                        decode_four_codes(out, codes);
+                        out += 1;
+                        break;
+                    case 3:
+                        codes[3] = 0;
+                        decode_four_codes(out, codes);
+                        out += 2;
+                        break;
+                }
+                *data_out_size = (size_t)out - (size_t)data_out;
```
Comment (benpicco) [inline sys/base64/base64.c:257]:
```suggestion
                *data_out_size = (uintptr_t)out - (uintptr_t)data_out;
```

makes no difference, but easier to understand.

File: sys/base64/base64.c [hunk 5]
```diff
@@ -232,34 +229,44 @@ int base64_decode(const void *base64_in, size_t base64_in_size,
...
-        code = getcode(in[i]);
-
-        if (code == BASE64_NOT_DEFINED || code == BASE64_EQUALS) {
-            continue;
+    const uint8_t *end = base64_in + base64_in_size;
+    uint8_t codes[4];
+    size_t n_codes;
+
+    while (1) {
+        n_codes = 0;
```
Comment (benpicco) [inline sys/base64/base64.c:237]:
```suggestion

    while (1) {
        size_t n_codes = 0;
```

**Comment by benpicco** [comment]
Please squash!
I guess we should wait till after Friday with merging this.

**Comment by miri64** [comment]
> I guess we should wait till after Friday with merging this.

Why? I wouldn't consider a well tested application level en-/decoder cleanup as high impact.

**Comment by maribu** [comment]
Thanks for the reviews!

**Pull request merged**